BUG: optimize: test failures in scikit-learn following LBFGS C rewrite

### Describe your issue.

We notice some tests failures in scikit-learn in our CI build testing against scipy development wheels from scientific-python-nightly-wheels https://github.com/scikit-learn/scikit-learn/issues/29864.

This seems related to LBFGS C rewrite https://github.com/scipy/scipy/pull/21314. Below is a snippet that raises an `AssertionError` on 97445cc60 (merge commit for LBGFS rewrite) and passes on 4bd5e943c (last commit on `main` before LBFGS rewrite)

I extracted the snippet from our scikit-learn tests, I am sure with more time it could be simplified a lot more ...

To give some context, this compare the coefficients obtained from two different solvers for [Ridge regression](https://scikit-learn.org/stable/modules/linear_model.html#regression), one being with scipy LBFGS.

### Reproducing Code Example

```python
import warnings

import numpy as np
from numpy.testing import assert_allclose

from scipy import optimize


def make_regression(
    n_samples=100,
    n_features=100,
    random_state=None,
):
    n_informative = 10
    n_targets = 1
    generator = np.random.RandomState(random_state)
    X = generator.standard_normal(size=(n_samples, n_features))

    # Generate a ground truth model with only n_informative features being non
    # zeros (the other features are not correlated to y and should be ignored
    # by a sparsifying regularizers such as L1 or elastic net)
    ground_truth = np.zeros((n_features, n_targets))
    ground_truth[:n_informative, :] = 100 * generator.uniform(
        size=(n_informative, n_targets)
    )

    y = np.dot(X, ground_truth)
    y = np.squeeze(y)

    return X, y

def _solve_svd(X, y, alpha):
    U, s, Vt = np.linalg.svd(X, full_matrices=False)
    idx = s > 1e-15  # same default value as scipy.linalg.pinv
    s_nnz = s[idx][:, None]
    UTy = U.T @ y
    d = np.zeros((s.shape[0], alpha.shape[0]), dtype=X.dtype)
    d[idx] = s_nnz / (s_nnz**2 + alpha)
    d_UT_y = d * UTy
    return (Vt.T @ d_UT_y).T


def _solve_lbfgs(
    X,
    y,
    alpha,
    positive=True,
    max_iter=None,
    tol=1e-4,
):
    """Solve ridge regression with LBFGS.

    The main purpose is fitting with forcing coefficients to be positive.
    For unconstrained ridge regression, there are faster dedicated solver methods.
    Note that with positive bounds on the coefficients, LBFGS seems faster
    than scipy.optimize.lsq_linear.
    """
    n_samples, n_features = X.shape

    options = {}
    if max_iter is not None:
        options["maxiter"] = max_iter
    config = {
        "method": "L-BFGS-B",
        "tol": tol,
        "jac": True,
        "options": options,
    }
    if positive:
        config["bounds"] = [(0, np.inf)] * n_features

    coefs = np.empty((y.shape[1], n_features), dtype=X.dtype)

    for i in range(y.shape[1]):
        x0 = np.zeros((n_features,))
        y_column = y[:, i]

        def func(w):
            residual = X.dot(w) - y_column
            f = 0.5 * residual.dot(residual) + 0.5 * alpha[i] * w.dot(w)
            grad = X.T @ residual + alpha[i] * w

            return f, grad

        result = optimize.minimize(func, x0, **config)
        if not result["success"]:
            warnings.warn(
                (
                    "The lbfgs solver did not converge. Try increasing max_iter "
                    f"or tol. Currently: max_iter={max_iter} and tol={tol}"
                ),
            )
        coefs[i] = result["x"]

    return coefs

alpha = 1e-3
X, y = make_regression(n_samples=300, n_features=300, random_state=42)
y = np.expand_dims(y, 1)
alpha = np.asarray([alpha])
config = {
    "positive": False,
    "tol": 1e-16,
    "max_iter": 500000,
}


coef_lbfgs = _solve_lbfgs(X, y, alpha, **config)
coef_cholesky = _solve_svd(X, y, alpha)
assert_allclose(coef_lbfgs, coef_cholesky, atol=1e-4, rtol=0)
```


### Error message

```shell
AssertionError: 
Not equal to tolerance rtol=0, atol=0.0001

Mismatched elements: 297 / 300 (99%)
Max absolute difference among violations: 0.09078348
Max relative difference among violations: 1.68720775
 ACTUAL: array([[ 3.538504e+01,  8.541181e+01,  8.063357e+01,  3.148950e+01,
         4.826480e+01,  4.467394e+01,  8.893688e+01,  6.085203e+01,
         4.375525e+01,  5.247418e+01,  2.049231e-01, -3.350157e-01,...
 DESIRED: array([[ 3.542202e+01,  8.542891e+01,  8.065989e+01,  3.144128e+01,
         4.823986e+01,  4.470257e+01,  8.896726e+01,  6.087696e+01,
         4.375691e+01,  5.245160e+01,  1.962697e-01, -3.143606e-01,...
```


### SciPy/NumPy/Python version and system information

```shell
Python 3.12.5 | packaged by conda-forge | (main, Aug  8 2024, 18:36:51) [GCC 12.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.show_config()
1.15.0.dev0+git20240917.4bd5e94 2.1.1 sys.version_info(major=3, minor=12, micro=5, releaselevel='final', serial=0)
/home/lesteve/dev/scipy/build-install/lib/python3.12/site-packages/scipy/__config__.py:154: UserWarning: Install `pyyaml` for better output
  warnings.warn("Install `pyyaml` for better output", stacklevel=1)
{
  "Compilers": {
    "c": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "14.2.1",
      "commands": "ccache, cc"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "3.0.11",
      "commands": "cython"
    },
    "c++": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "14.2.1",
      "commands": "ccache, c++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld.bfd",
      "version": "14.2.1",
      "commands": "gfortran"
    },
    "pythran": {
      "version": "0.16.1",
      "include directory": "../../../micromamba/envs/scipy-dev/lib/python3.12/site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "linux"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "openblas",
      "found": true,
      "version": "0.3.28",
      "detection method": "pkgconfig",
      "include directory": "/usr/include/openblas",
      "lib directory": "/usr/lib",
      "openblas configuration": "USE_64BITINT= NO_CBLAS= NO_LAPACK= NO_LAPACKE= DYNAMIC_ARCH=ON DYNAMIC_OLDER=OFF NO_AFFINITY=ON USE_OPENMP=1 CORE2 MAX_THREADS=64",
      "pc file directory": "/usr/lib/pkgconfig"
    },
    "lapack": {
      "name": "openblas",
      "found": true,
      "version": "0.3.28",
      "detection method": "pkgconfig",
      "include directory": "/usr/include/openblas",
      "lib directory": "/usr/lib",
      "openblas configuration": "USE_64BITINT= NO_CBLAS= NO_LAPACK= NO_LAPACKE= DYNAMIC_ARCH=ON DYNAMIC_OLDER=OFF NO_AFFINITY=ON USE_OPENMP=1 CORE2 MAX_THREADS=64",
      "pc file directory": "/usr/lib/pkgconfig"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.13.6",
      "detection method": "config-tool",
      "include directory": "unknown"
    }
  },
  "Python Information": {
    "path": "/home/lesteve/micromamba/envs/scipy-dev/bin/python3.12",
    "version": "3.12"
  }
}
```

--- a/scipy/optimize/__lbfgsb.c
+++ b/scipy/optimize/__lbfgsb.c
@@ -102,8 +102,6 @@
  *
  * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
 
-
-
 #include "__lbfgsb.h"
 
 
@@ -1589,7 +1587,7 @@ cauchy(int n, double* x, double* l, double* u,
             for (j = 0; j < col; j++) {
                 p[j] = p[j] + wy[i + n*pointr]*neggi;
                 p[col + j] = p[col + j] + ws[i + n*pointr]*neggi;
-                pointr = (pointr % m) + 1;
+                pointr = (pointr + 1) % m;
             }
             // 40
             if ((nbd[i] <= 2) && (nbd[i] != 0) && (neggi < 0.0))
@@ -1737,7 +1735,7 @@ cauchy(int n, double* x, double* l, double* u,
             for (j = 0; j < col; j++) {
                 wbp[j] = wy[ibp + n*pointr];
                 wbp[col + j] = theta * ws[ibp + n*pointr];
-                pointr = (pointr % m) + 1;
+                pointr = (pointr + 1) % m;
             }
             // 70
 
@@ -1851,7 +1849,7 @@ cmprlb(int n, int m, double* x, double* g,
                 r[i] = r[i] + wy[k + n*pointr]*a1 + ws[k + n*pointr]*a2;
             }
             // 32
-            pointr = (pointr % m) + 1;
+            pointr = (pointr + 1) % m;
         }
         // 34
     }
@@ -2110,7 +2108,7 @@ formk(int n, int nsub, int* ind, int nenter, int ileave,
             wn1[iy + 2*m*jy] = temp1;
             wn1[is + 2*m*js] = temp2;
             wn1[is + 2*m*jy] = temp3;
-            jpntr = (jpntr % m) + 1;
+            jpntr = (jpntr + 1) % m;
         }
         // 20
 
@@ -2131,7 +2129,7 @@ formk(int n, int nsub, int* ind, int nenter, int ileave,
                 temp3 = temp3 + ws[k1 + n*ipntr]*wy[k1 + n*jpntr];
             }
             // 25
-            ipntr = (ipntr % m) + 1;
+            ipntr = (ipntr + 1) % m;
             wn1[is + 2*m*jy] = temp3;
         }
         // 30
@@ -2173,11 +2171,11 @@ formk(int n, int nsub, int* ind, int nenter, int ileave,
 
             wn1[iy + 2*m*jy] = wn1[iy + 2*m*jy] + temp1 - temp3;
             wn1[is + 2*m*js] = wn1[is + 2*m*js] - temp2 + temp4;
-            jpntr = (jpntr % m) + 1;
+            jpntr = (jpntr + 1) % m;
         }
         // 40
 
-        ipntr = (ipntr % m) + 1;
+        ipntr = (ipntr + 1) % m;
     }
     // 45
 
@@ -2208,10 +2206,10 @@ formk(int n, int nsub, int* ind, int nenter, int ileave,
             } else {
                 wn1[is + 2*m*jy] = wn1[is + 2*m*jy] - temp1 + temp3;
             }
-            jpntr = (jpntr % m) + 1;
+            jpntr = (jpntr + 1) % m;
         }
         // 55
-        ipntr = (ipntr % m) + 1;
+        ipntr = (ipntr + 1) % m;
     }
     // 60
 
@@ -2586,16 +2584,11 @@ lnsrlb(int n, double* l, double* u, int* nbd, double* x,
     //       to perform the line search.  Subroutine dscrch is safeguarded so
     //       that all trial points lie within the feasible region.
     //
-    //     Be mindful that the dcsrch subroutine being called is a copy in
-    //       this file (lbfgsb.f) and NOT in the Minpack2 copy distributed
-    //       by scipy.
-    //
     //     Subprograms called:
     //
     //       Minpack2 Library ... dcsrch.
     //
     //       Linpack ... dtrsl, ddot.
-    //
     //
     //                           *  *  *
     //
@@ -2753,8 +2746,8 @@ matupd(int n, int m, double* ws, double *wy, double* sy, double* ss,
         *col = iupdat;
         *itail = (*head + iupdat - 1) % m;
     } else {
-        *itail = (*itail % m);
-        *head = (*head % m);
+        *itail = (*itail + 1) % m;
+        *head = (*head + 1) % m;
     }
 
     // Update matrices WS and WY.
@@ -2783,7 +2776,7 @@ matupd(int n, int m, double* ws, double *wy, double* sy, double* ss,
     {
         sy[*col - 1 + m*j] = ddot_(&n, d, &one_int, &wy[pointr*n], &one_int);
         ss[j + m*(*col - 1)] = ddot_(&n, &ws[pointr*n], &one_int, &d[0], &one_int);
-        pointr = (pointr % m) + 1;
+        pointr = (pointr + 1) % m;
     }
     // 51
 
@@ -3051,7 +3044,7 @@ void subsm(int n, int m, int nsub, int* ind,
         }
         wv[i] = temp1;
         wv[col + i] = theta*temp2;
-        pointr = (pointr % m) + 1;
+        pointr = (pointr + 1) % m;
     }
 
     // Compute wv:=K^(-1)wv.
@@ -3080,7 +3073,7 @@ void subsm(int n, int m, int nsub, int* ind,
                           (ws[k + n*pointr] * wv[js]);
         }
         // 30
-        pointr = (pointr % m) + 1;
+        pointr = (pointr + 1) % m;
     }
     // 40
 

--- a/scipy/optimize/__lbfgsb.h
+++ b/scipy/optimize/__lbfgsb.h
@@ -9,8 +9,6 @@ extern "C"
 #define PY_SSIZE_T_CLEAN
 #include "Python.h"
 #include "numpy/arrayobject.h"
-#include "../_build_utils/src/npy_cblas.h"
-#include "../_build_utils/src/fortran_defs.h"
 #include <math.h>
 
 #define PYERR(errobj,message) {PyErr_SetString(errobj,message); return NULL;}

